node/src/neighborhood/overall_connection_status.rs

```diff
@@ -0,0 +1,960 @@
+use masq_lib::ui_gateway::{MessageTarget, NodeToUiMessage};
+use std::net::IpAddr;
+
+#[derive(PartialEq, Debug, Clone)]
+pub enum ConnectionStageErrors {
+    TcpConnectionFailed,
+    NoGossipResponseReceived,
+    DeadEndFound,
+}
+
+#[derive(PartialEq, Debug, Clone)]
+pub enum ConnectionStage {
+    StageZero,
+    TcpConnectionEstablished,
+    NeighborshipEstablished,
+    Failed(ConnectionStageErrors),
+}
+
+impl TryFrom<&ConnectionStage> for usize {
+    type Error = ();
+
+    fn try_from(connection_stage: &ConnectionStage) -> Result<Self, Self::Error> {
+        match connection_stage {
+            ConnectionStage::StageZero => Ok(0),
+            ConnectionStage::TcpConnectionEstablished => Ok(1),
+            ConnectionStage::NeighborshipEstablished => Ok(2),
+            ConnectionStage::Failed(_) => Err(()),
+        }
+    }
+}
+
+#[derive(PartialEq, Debug, Clone)]
+pub struct ConnectionProgress {
+    pub initial_node_descriptor: NodeDescriptor,
+    pub current_peer_addr: IpAddr,
+    pub connection_stage: ConnectionStage,
+}
+
+impl ConnectionProgress {
+    pub fn new(node_descriptor: &NodeDescriptor) -> Self {
+        let node_addr = node_descriptor.node_addr_opt.as_ref().unwrap_or_else(|| {
+            panic!(
+                "Unable to receive node addr for the descriptor {:?}",
+                node_descriptor
+            )
+        });
+        Self {
+            initial_node_descriptor: node_descriptor.clone(),
+            current_peer_addr: node_addr.ip_addr(),
+            connection_stage: ConnectionStage::StageZero,
+        }
+    }
+
+    pub fn update_stage(&mut self, connection_stage: ConnectionStage) {
+        let current_stage = usize::try_from(&self.connection_stage);
+        let new_stage = usize::try_from(&connection_stage);
+
+        if let (Ok(current_stage_num), Ok(new_stage_num)) = (current_stage, new_stage) {
+            if new_stage_num != current_stage_num + 1 {
+                panic!(
+                    "Can't update the stage from {:?} to {:?}",
+                    self.connection_stage, connection_stage
+                )
+            }
+        }
+
+        self.connection_stage = connection_stage;
+    }
+
+    pub fn handle_pass_gossip(&mut self, new_pass_target: IpAddr) {
+        if self.connection_stage != ConnectionStage::TcpConnectionEstablished {
+            panic!(
+                "Can't update the stage from {:?} to {:?}",
+                self.connection_stage,
+                ConnectionStage::StageZero
+            )
+        };
+
+        self.connection_stage = ConnectionStage::StageZero;
+        self.current_peer_addr = new_pass_target;
+    }
+}
+
+#[derive(PartialEq, Debug, Copy, Clone)]
+enum OverallConnectionStage {
+    NotConnected = 0,        // Not connected to any neighbor.
+    ConnectedToNeighbor = 1, // Neighborship established. Same as No 3 hops route found.
+    ThreeHopsRouteFound = 2, // check_connectedness() returned true, data can now be relayed.
```

Comment:
I searched for it, and every solution involves to explicitly enter the numbers.

Right now, we have a similar test, and I think since we only have three stages, this test is good enough.

```rust
    #[test]
    fn overall_connection_stage_can_be_converted_into_usize_and_can_be_compared() {
        assert!(
            OverallConnectionStage::ConnectedToNeighbor as usize
                > OverallConnectionStage::NotConnected as usize
        );
        assert!(
            OverallConnectionStage::ThreeHopsRouteFound as usize
                > OverallConnectionStage::ConnectedToNeighbor as usize
        );
    }
```